From 2fb3e0eb753fd8cd3b21bfd122e807212d17f316 Mon Sep 17 00:00:00 2001
From: Lionel Landwerlin <lionel.g.landwerlin@linux.intel.com>
Date: Mon, 10 Oct 2011 11:13:55 +0100
Subject: [PATCH] Don't lose application redraws

Our usage of DamageReportBoundingBox was causing us to miss some
updates when an area of the screen was drawn twice in rapid
succession. Add an explicit XSync() call to force the server
to flush rendering to the kernel before we draw.

https://bugzilla.gnome.org/show_bug.cgi?id=657071

Patch from Owen Taylor
---
 src/compositor/mutter-window.c |   21 +++++++++++++++++++++
 1 files changed, 21 insertions(+), 0 deletions(-)

diff --git a/src/compositor/mutter-window.c b/src/compositor/mutter-window.c
index 5026021..cee6b50 100644
--- a/src/compositor/mutter-window.c
+++ b/src/compositor/mutter-window.c
@@ -1833,6 +1833,27 @@ mutter_window_pre_paint (MutterWindow *self)
   if (priv->received_damage)
     {
       XDamageSubtract (xdisplay, priv->damage, None, None);
+
+      /* We need to make sure that any X drawing that happens before the
+       * XDamageSubtract() above is visible to subsequent GL rendering;
+       * the only standardized way to do this is EXT_x11_sync_object,
+       * which isn't yet widely available. For now, we count on details
+       * of Xorg and the open source drivers, and hope for the best
+       * otherwise.
+       *
+       * Xorg and open source driver specifics:
+       *
+       * The X server makes sure to flush drawing to the kernel before
+       * sending out damage events, but since we use DamageReportBoundingBox
+       * there may be drawing between the last damage event and the
+       * XDamageSubtract() that needs to be flushed as well.
+       *
+       * Xorg always makes sure that drawing is flushed to the kernel
+       * before writing events or responses to the client, so any round trip
+       * request at this point is sufficient to flush the GLX buffers.
+       */
+      XSync (xdisplay, False);
+
       priv->received_damage = FALSE;
     }
 
-- 
1.7.6.3

